src/Sulu/Bundle/AdminBundle/Resources/js/components/AutoComplete/AutoComplete.js

```diff
@@ -0,0 +1,182 @@
+    loading?: boolean,
+    /** Called on every input change */
+    onChange: (value: string) => void,
+    /** Debounced version of `onChange`. Put network requests inside of this handler */
+    onDebouncedChange: (value: string) => void,
+    onSuggestionSelection: (value: string) => void,
+    /** Message when no `Suggestions` could be found */
+    noSuggestionsMessage?: string,
+};
+
+@observer
+export default class AutoComplete extends React.PureComponent<Props> {
+    static defaultProps = {
+        value: '',
+        threshold: 0,
+    };
+
+    static Suggestion = Suggestion;
+
+    @observable open: boolean = false;
+
+    @observable inputRef: ElementRef<*>;
+
+    previousSuggestionListChildrenCount: number = 0;
+
+    componentWillUnmount() {
+        this.debouncedChange.clear();
+    }
+
+    @action openSuggestions() {
+        this.open = true;
+    }
+
+    @action closeSuggestions() {
+        this.open = false;
+    }
+
+    @computed get suggestionStyle(): Object {
+        const suggestionListMinWidth = (this.inputRef) ? this.inputRef.scrollWidth - POPOVER_HORIZONTAL_OFFSET * 2: 0;
+
+        return {
+            minWidth: suggestionListMinWidth,
+        };
+    }
+
+    createSuggestions(children: ChildrenArray<Element<typeof Suggestion>>) {
+        const {value} = this.props;
+
+        return React.Children.map(children, (child, index: number) => {
+            return (
+                <li
+                    style={this.suggestionStyle}
+                    className={autoCompleteStyles.suggestionItem}
+                >
+                    {
+                        React.cloneElement(child, {
+                            key: index,
+                            query: value,
+                            onSelection: this.handleSuggestionSelection,
+                        })
+                    }
+                </li>
+            );
+        });
+    }
+
+    createNoSuggestionsMessage(): ElementRef<'li'> {
+        const {noSuggestionsMessage} = this.props;
+
+        if (!noSuggestionsMessage) {
+            return null;
+        }
+
+        return (
+            <li
+                style={this.suggestionStyle}
+                className={autoCompleteStyles.noSuggestions}
+            >
+                {noSuggestionsMessage}
+            </li>
+        );
+    }
+
+    setInputRef = (inputRef: ElementRef<'label'>) => {
+        if (inputRef) {
+            this.inputRef = inputRef;
+        }
+    };
+
+    debouncedChange = debounce((value: string) => {
+        this.props.onDebouncedChange(value);
+    }, DEBOUNCE_TIME);
+
+    handleSuggestionSelection = (value: string) => {
+        this.props.onSuggestionSelection(value);
+        this.closeSuggestions();
+    };
+
+    handleChange = (value: string) => {
+        if (value !== '') {
+            this.openSuggestions();
+        } else {
+            this.closeSuggestions();
+        }
+
+        this.props.onChange(value);
```

Comment:
As already written before, I think this value should be only passed to the `onChange` handler if it is an available option. That probably means that one of the three available callbacks is obsolete. Let's talk about that on monday.